pkg/apis/admission/types.go

```diff
@@ -0,0 +1,70 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/kubernetes/pkg/apis/authentication"
+)
+
+// AdmittanceReview describes an admission request.
+type AdmittanceReview struct {
+	metav1.TypeMeta
+
+	// Spec describes the attributes for the admission request.
+	Spec AdmittanceReviewSpec
+	// Status is filled in by the webhook and indicates whether the admission request should be permitted.
+	Status AdmittanceReviewStatus
+}
+
+// AdmittanceReviewSpec describes the admission.Attributes for the admission request.
+type AdmittanceReviewSpec struct {
+	// Kind is the type of object being manipulated.  For example: Pod
+	Kind metav1.GroupVersionKind
+	// Name is the name of the object as presented in the request.  On a CREATE operation, the client may omit name and
+	// rely on the server to generate the name.  If that is the case, this method will return the empty string.
+	Name string
+	// Namespace is the namespace associated with the request (if any).
+	Namespace string
+	// Object is the object from the incoming request prior to default values being applied
+	Object runtime.Object
+	// OldObject is the existing object. Only populated for UPDATE requests.
+	OldObject runtime.Object
+	// Operation is the operation being performed
+	Operation admission.Operation
+	// Resource is the name of the resource being requested.  This is not the kind.  For example: pods
+	Resource metav1.GroupVersionResource
+	// SubResource is the name of the subresource being requested.  This is a different resource, scoped to the parent
+	// resource, but it may have a different kind. For instance, /pods has the resource "pods" and the kind "Pod", while
+	// /pods/foo/status has the resource "pods", the sub resource "status", and the kind "Pod" (because status operates on
+	// pods). The binding resource for a pod though may be /pods/foo/binding, which has resource "pods", subresource
+	// "binding", and kind "Binding".
+	SubResource string
+	// UserInfo is information about the requesting user
+	UserInfo authentication.UserInfo
+}
+
+// AdmittanceReviewStatus describes the status of the admission request.
+type AdmittanceReviewStatus struct {
+	// Allowed indicates whether or not the admission request was permitted.
+	Allowed bool
+	// Reason should be empty unless Allowed is false in which case it may contain a short description of what is wrong.
+	// Kubernetes may truncate excessively long errors when displaying to the user.
+	Reason string
```

Comment:
This should be Message, not Reason (by description).  You want to have both `Message` and `Reason`, and use the same descriptions from `metav1.Status` (i.e. Reason is a constant, machine readable, Message is a description, human readable).  And preserve the part about truncation, that is important.